{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Import libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "from copy import deepcopy"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Read the sample data and see what it looks like"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>study_id</th>\n",
       "      <th>specimen collector sample ID</th>\n",
       "      <th>sample collected by</th>\n",
       "      <th>sequence submitted by</th>\n",
       "      <th>sample collection date</th>\n",
       "      <th>sample collection date null reason</th>\n",
       "      <th>geo_loc_name (country)</th>\n",
       "      <th>geo_loc_name (state/province/territory)</th>\n",
       "      <th>organism</th>\n",
       "      <th>isolate</th>\n",
       "      <th>...</th>\n",
       "      <th>consensus sequence software name</th>\n",
       "      <th>consensus sequence software version</th>\n",
       "      <th>breadth of coverage value</th>\n",
       "      <th>depth of coverage value</th>\n",
       "      <th>reference genome accession</th>\n",
       "      <th>bioinformatics protocol</th>\n",
       "      <th>gene name</th>\n",
       "      <th>diagnostic pcr Ct value</th>\n",
       "      <th>diagnostic pcr Ct value null reason</th>\n",
       "      <th>GISAID accession</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>UHTC-ON</td>\n",
       "      <td>UHTC_0471</td>\n",
       "      <td>Unity Health Toronto</td>\n",
       "      <td>Ontario Institute for Cancer Research (OICR)</td>\n",
       "      <td>2020-12-15</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Canada</td>\n",
       "      <td>Ontario</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>hCoV-19/Canada/ON-UHTC_0471/2021</td>\n",
       "      <td>...</td>\n",
       "      <td>ARTIC-nanopolish</td>\n",
       "      <td>1.1.3</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>5384.4X</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>UHTC-ON</td>\n",
       "      <td>UHTC_0464</td>\n",
       "      <td>Unity Health Toronto</td>\n",
       "      <td>Ontario Institute for Cancer Research (OICR)</td>\n",
       "      <td>2020-10-12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Canada</td>\n",
       "      <td>Ontario</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>hCoV-19/Canada/ON-UHTC_0464/2021</td>\n",
       "      <td>...</td>\n",
       "      <td>ARTIC-nanopolish</td>\n",
       "      <td>1.1.3</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>2912.4X</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>UHTC-ON</td>\n",
       "      <td>UHTC_0463</td>\n",
       "      <td>Unity Health Toronto</td>\n",
       "      <td>Ontario Institute for Cancer Research (OICR)</td>\n",
       "      <td>2020-10-02</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Canada</td>\n",
       "      <td>Ontario</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>hCoV-19/Canada/ON-UHTC_0463/2021</td>\n",
       "      <td>...</td>\n",
       "      <td>ARTIC-nanopolish</td>\n",
       "      <td>1.1.3</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>2661.6X</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>KHSC-ON</td>\n",
       "      <td>SLB3076</td>\n",
       "      <td>Queen's University / Kingston Health Sciences ...</td>\n",
       "      <td>Ontario Institute for Cancer Research (OICR)</td>\n",
       "      <td>2020-04-12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Canada</td>\n",
       "      <td>Ontario</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>hCoV-19/Canada/ON-SLB3076/2021</td>\n",
       "      <td>...</td>\n",
       "      <td>ncov2019-artic-nf</td>\n",
       "      <td>OICR v1.6</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>1671.9X</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>KHSC-ON</td>\n",
       "      <td>SLB3121</td>\n",
       "      <td>Queen's University / Kingston Health Sciences ...</td>\n",
       "      <td>Ontario Institute for Cancer Research (OICR)</td>\n",
       "      <td>2020-04-22</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Canada</td>\n",
       "      <td>Ontario</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>hCoV-19/Canada/ON-SLB3121/2021</td>\n",
       "      <td>...</td>\n",
       "      <td>ncov2019-artic-nf</td>\n",
       "      <td>OICR v1.6</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>959.3X</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 43 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  study_id specimen collector sample ID  \\\n",
       "0  UHTC-ON                    UHTC_0471   \n",
       "1  UHTC-ON                    UHTC_0464   \n",
       "2  UHTC-ON                    UHTC_0463   \n",
       "3  KHSC-ON                      SLB3076   \n",
       "4  KHSC-ON                      SLB3121   \n",
       "\n",
       "                                 sample collected by  \\\n",
       "0                               Unity Health Toronto   \n",
       "1                               Unity Health Toronto   \n",
       "2                               Unity Health Toronto   \n",
       "3  Queen's University / Kingston Health Sciences ...   \n",
       "4  Queen's University / Kingston Health Sciences ...   \n",
       "\n",
       "                          sequence submitted by sample collection date  \\\n",
       "0  Ontario Institute for Cancer Research (OICR)             2020-12-15   \n",
       "1  Ontario Institute for Cancer Research (OICR)             2020-10-12   \n",
       "2  Ontario Institute for Cancer Research (OICR)             2020-10-02   \n",
       "3  Ontario Institute for Cancer Research (OICR)             2020-04-12   \n",
       "4  Ontario Institute for Cancer Research (OICR)             2020-04-22   \n",
       "\n",
       "   sample collection date null reason geo_loc_name (country)  \\\n",
       "0                                 NaN                 Canada   \n",
       "1                                 NaN                 Canada   \n",
       "2                                 NaN                 Canada   \n",
       "3                                 NaN                 Canada   \n",
       "4                                 NaN                 Canada   \n",
       "\n",
       "  geo_loc_name (state/province/territory)      organism  \\\n",
       "0                                 Ontario  Not Provided   \n",
       "1                                 Ontario  Not Provided   \n",
       "2                                 Ontario  Not Provided   \n",
       "3                                 Ontario  Not Provided   \n",
       "4                                 Ontario  Not Provided   \n",
       "\n",
       "                            isolate  ... consensus sequence software name  \\\n",
       "0  hCoV-19/Canada/ON-UHTC_0471/2021  ...                 ARTIC-nanopolish   \n",
       "1  hCoV-19/Canada/ON-UHTC_0464/2021  ...                 ARTIC-nanopolish   \n",
       "2  hCoV-19/Canada/ON-UHTC_0463/2021  ...                 ARTIC-nanopolish   \n",
       "3    hCoV-19/Canada/ON-SLB3076/2021  ...                ncov2019-artic-nf   \n",
       "4    hCoV-19/Canada/ON-SLB3121/2021  ...                ncov2019-artic-nf   \n",
       "\n",
       "  consensus sequence software version breadth of coverage value  \\\n",
       "0                               1.1.3              Not Provided   \n",
       "1                               1.1.3              Not Provided   \n",
       "2                               1.1.3              Not Provided   \n",
       "3                           OICR v1.6              Not Provided   \n",
       "4                           OICR v1.6              Not Provided   \n",
       "\n",
       "  depth of coverage value reference genome accession bioinformatics protocol  \\\n",
       "0                 5384.4X                        NaN                     NaN   \n",
       "1                 2912.4X                        NaN                     NaN   \n",
       "2                 2661.6X                        NaN                     NaN   \n",
       "3                 1671.9X                        NaN                     NaN   \n",
       "4                  959.3X                        NaN                     NaN   \n",
       "\n",
       "      gene name diagnostic pcr Ct value diagnostic pcr Ct value null reason  \\\n",
       "0  Not Provided                     NaN                        Not Provided   \n",
       "1  Not Provided                     NaN                        Not Provided   \n",
       "2  Not Provided                     NaN                        Not Provided   \n",
       "3  Not Provided                     NaN                        Not Provided   \n",
       "4  Not Provided                     NaN                        Not Provided   \n",
       "\n",
       "  GISAID accession  \n",
       "0              NaN  \n",
       "1              NaN  \n",
       "2              NaN  \n",
       "3              NaN  \n",
       "4              NaN  \n",
       "\n",
       "[5 rows x 43 columns]"
      ]
     },
     "execution_count": 114,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seqs = pd.read_csv(\"data/files-archive.tsv\", sep = \"\\t\")\n",
    "seqs.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Print out all column names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['study_id', 'specimen collector sample ID', 'sample collected by',\n",
      "       'sequence submitted by', 'sample collection date',\n",
      "       'sample collection date null reason', 'geo_loc_name (country)',\n",
      "       'geo_loc_name (state/province/territory)', 'organism', 'isolate',\n",
      "       'fasta header name', 'purpose of sampling',\n",
      "       'purpose of sampling details', 'anatomical material', 'anatomical part',\n",
      "       'body product', 'environmental material', 'environmental site',\n",
      "       'collection device', 'collection method', 'host (scientific name)',\n",
      "       'host disease', 'host age', 'host age null reason', 'host age unit',\n",
      "       'host age bin', 'host gender', 'purpose of sequencing',\n",
      "       'purpose of sequencing details', 'sequencing instrument',\n",
      "       'sequencing protocol', 'raw sequence data processing method',\n",
      "       'dehosting method', 'consensus sequence software name',\n",
      "       'consensus sequence software version', 'breadth of coverage value',\n",
      "       'depth of coverage value', 'reference genome accession',\n",
      "       'bioinformatics protocol', 'gene name', 'diagnostic pcr Ct value',\n",
      "       'diagnostic pcr Ct value null reason', 'GISAID accession'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "print(seqs.columns)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Remove the columns we don't need"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [],
   "source": [
    "seqs.drop(columns=['study_id', 'specimen collector sample ID', 'sample collected by',\n",
    "       'sequence submitted by',\n",
    "       'sample collection date null reason', 'geo_loc_name (country)',\n",
    "       'organism', 'isolate',\n",
    "       'fasta header name',\n",
    "       'purpose of sampling details', 'anatomical material', 'anatomical part',\n",
    "       'body product', 'environmental material', 'environmental site',\n",
    "       'collection device', 'collection method', 'host (scientific name)',\n",
    "       'host disease','host age null reason', 'host age unit',\n",
    "       'purpose of sequencing details',\n",
    "       'sequencing protocol', 'raw sequence data processing method',\n",
    "       'dehosting method', 'consensus sequence software name',\n",
    "       'consensus sequence software version', 'reference genome accession',\n",
    "       'bioinformatics protocol', 'gene name', \n",
    "       'diagnostic pcr Ct value null reason'], inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample collection date</th>\n",
       "      <th>geo_loc_name (state/province/territory)</th>\n",
       "      <th>purpose of sampling</th>\n",
       "      <th>host age</th>\n",
       "      <th>host age bin</th>\n",
       "      <th>host gender</th>\n",
       "      <th>purpose of sequencing</th>\n",
       "      <th>sequencing instrument</th>\n",
       "      <th>breadth of coverage value</th>\n",
       "      <th>depth of coverage value</th>\n",
       "      <th>diagnostic pcr Ct value</th>\n",
       "      <th>GISAID accession</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2020-12-15</td>\n",
       "      <td>Ontario</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>Oxford Nanopore</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>5384.4X</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2020-10-12</td>\n",
       "      <td>Ontario</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>Oxford Nanopore</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>2912.4X</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2020-10-02</td>\n",
       "      <td>Ontario</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>Oxford Nanopore</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>2661.6X</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2020-04-12</td>\n",
       "      <td>Ontario</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>Illumina</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>1671.9X</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2020-04-22</td>\n",
       "      <td>Ontario</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>Illumina</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>959.3X</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample collection date geo_loc_name (state/province/territory)  \\\n",
       "0             2020-12-15                                 Ontario   \n",
       "1             2020-10-12                                 Ontario   \n",
       "2             2020-10-02                                 Ontario   \n",
       "3             2020-04-12                                 Ontario   \n",
       "4             2020-04-22                                 Ontario   \n",
       "\n",
       "  purpose of sampling  host age    host age bin   host gender  \\\n",
       "0        Not Provided       NaN  Not Applicable  Not Provided   \n",
       "1        Not Provided       NaN  Not Applicable  Not Provided   \n",
       "2        Not Provided       NaN  Not Applicable  Not Provided   \n",
       "3        Not Provided       NaN  Not Applicable  Not Provided   \n",
       "4        Not Provided       NaN  Not Applicable  Not Provided   \n",
       "\n",
       "  purpose of sequencing sequencing instrument breadth of coverage value  \\\n",
       "0          Not Provided       Oxford Nanopore              Not Provided   \n",
       "1          Not Provided       Oxford Nanopore              Not Provided   \n",
       "2          Not Provided       Oxford Nanopore              Not Provided   \n",
       "3          Not Provided              Illumina              Not Provided   \n",
       "4          Not Provided              Illumina              Not Provided   \n",
       "\n",
       "  depth of coverage value  diagnostic pcr Ct value GISAID accession  \n",
       "0                 5384.4X                      NaN              NaN  \n",
       "1                 2912.4X                      NaN              NaN  \n",
       "2                 2661.6X                      NaN              NaN  \n",
       "3                 1671.9X                      NaN              NaN  \n",
       "4                  959.3X                      NaN              NaN  "
      ]
     },
     "execution_count": 117,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seqs.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Let's replace the many forms of 'Not Provided' or NaN with np.nan and convert collection date to a date time object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample collection date</th>\n",
       "      <th>geo_loc_name (state/province/territory)</th>\n",
       "      <th>purpose of sampling</th>\n",
       "      <th>host age</th>\n",
       "      <th>host age bin</th>\n",
       "      <th>host gender</th>\n",
       "      <th>purpose of sequencing</th>\n",
       "      <th>sequencing instrument</th>\n",
       "      <th>breadth of coverage value</th>\n",
       "      <th>depth of coverage value</th>\n",
       "      <th>diagnostic pcr Ct value</th>\n",
       "      <th>GISAID accession</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2020-12-15</td>\n",
       "      <td>Ontario</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Oxford Nanopore</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5384.4X</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2020-10-12</td>\n",
       "      <td>Ontario</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Oxford Nanopore</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2912.4X</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2020-10-02</td>\n",
       "      <td>Ontario</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Oxford Nanopore</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2661.6X</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2020-04-12</td>\n",
       "      <td>Ontario</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Illumina</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1671.9X</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2020-04-22</td>\n",
       "      <td>Ontario</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Illumina</td>\n",
       "      <td>NaN</td>\n",
       "      <td>959.3X</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample collection date geo_loc_name (state/province/territory)  \\\n",
       "0             2020-12-15                                 Ontario   \n",
       "1             2020-10-12                                 Ontario   \n",
       "2             2020-10-02                                 Ontario   \n",
       "3             2020-04-12                                 Ontario   \n",
       "4             2020-04-22                                 Ontario   \n",
       "\n",
       "  purpose of sampling  host age host age bin host gender  \\\n",
       "0                 NaN       NaN          NaN         NaN   \n",
       "1                 NaN       NaN          NaN         NaN   \n",
       "2                 NaN       NaN          NaN         NaN   \n",
       "3                 NaN       NaN          NaN         NaN   \n",
       "4                 NaN       NaN          NaN         NaN   \n",
       "\n",
       "  purpose of sequencing sequencing instrument breadth of coverage value  \\\n",
       "0                   NaN       Oxford Nanopore                       NaN   \n",
       "1                   NaN       Oxford Nanopore                       NaN   \n",
       "2                   NaN       Oxford Nanopore                       NaN   \n",
       "3                   NaN              Illumina                       NaN   \n",
       "4                   NaN              Illumina                       NaN   \n",
       "\n",
       "  depth of coverage value  diagnostic pcr Ct value GISAID accession  \n",
       "0                 5384.4X                      NaN              NaN  \n",
       "1                 2912.4X                      NaN              NaN  \n",
       "2                 2661.6X                      NaN              NaN  \n",
       "3                 1671.9X                      NaN              NaN  \n",
       "4                  959.3X                      NaN              NaN  "
      ]
     },
     "execution_count": 118,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seqs.replace({'Not Provided':np.nan, 'NaN':np.nan, 'Not Applicable':np.nan}, inplace = True)\n",
    "seqs['sample collection date'] = pd.to_datetime(seqs['sample collection date'])\n",
    "seqs.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### What percent of the data is missing?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "sample collection date                      0.000000\n",
       "geo_loc_name (state/province/territory)     0.000000\n",
       "purpose of sampling                        21.533453\n",
       "host age                                   92.793075\n",
       "host age bin                               54.541226\n",
       "host gender                                56.600890\n",
       "purpose of sequencing                      22.678132\n",
       "sequencing instrument                       0.000000\n",
       "breadth of coverage value                   0.105356\n",
       "depth of coverage value                    25.508509\n",
       "diagnostic pcr Ct value                    73.663829\n",
       "GISAID accession                            1.722714\n",
       "dtype: float64"
      ]
     },
     "execution_count": 119,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seqs.isna().sum()/len(seqs)*100"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### All samples have associated collection dates!\n",
    "### Let's try grouping data by province to see if there's a pattern to what is missing."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>collection_date</th>\n",
       "      <th>sampling_purpose</th>\n",
       "      <th>age</th>\n",
       "      <th>age_bin</th>\n",
       "      <th>gender</th>\n",
       "      <th>sequencing_purpose</th>\n",
       "      <th>sequencing_instrument</th>\n",
       "      <th>coverage_breadth</th>\n",
       "      <th>coverage_depth</th>\n",
       "      <th>Ct</th>\n",
       "      <th>gisaid_accession</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>province</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Alberta</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>27740</td>\n",
       "      <td>27740</td>\n",
       "      <td>27740</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>26123</td>\n",
       "      <td>27740</td>\n",
       "      <td>1704</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>British Columbia</th>\n",
       "      <td>0</td>\n",
       "      <td>22576</td>\n",
       "      <td>34697</td>\n",
       "      <td>22576</td>\n",
       "      <td>22576</td>\n",
       "      <td>22576</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>34697</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Manitoba</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>25</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>124</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>New Brunswick</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>507</td>\n",
       "      <td>16</td>\n",
       "      <td>5</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>90</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Newfoundland and Labrador</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Nova Scotia</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1202</td>\n",
       "      <td>55</td>\n",
       "      <td>12</td>\n",
       "      <td>439</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>106</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ontario</th>\n",
       "      <td>0</td>\n",
       "      <td>111</td>\n",
       "      <td>29685</td>\n",
       "      <td>3229</td>\n",
       "      <td>5623</td>\n",
       "      <td>493</td>\n",
       "      <td>0</td>\n",
       "      <td>111</td>\n",
       "      <td>672</td>\n",
       "      <td>12022</td>\n",
       "      <td>111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Saskatchewan</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3907</td>\n",
       "      <td>3847</td>\n",
       "      <td>3677</td>\n",
       "      <td>344</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>80</td>\n",
       "      <td>2831</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           collection_date  sampling_purpose    age  age_bin  \\\n",
       "province                                                                       \n",
       "Alberta                                  0                 0  27740    27740   \n",
       "British Columbia                         0             22576  34697    22576   \n",
       "Manitoba                                 0                 0     22        0   \n",
       "New Brunswick                            0                 0    507       16   \n",
       "Newfoundland and Labrador                0                 0      4        0   \n",
       "Nova Scotia                              0                 0   1202       55   \n",
       "Ontario                                  0               111  29685     3229   \n",
       "Saskatchewan                             0                 0   3907     3847   \n",
       "\n",
       "                           gender  sequencing_purpose  sequencing_instrument  \\\n",
       "province                                                                       \n",
       "Alberta                     27740                   0                      0   \n",
       "British Columbia            22576               22576                      0   \n",
       "Manitoba                        0                  25                      0   \n",
       "New Brunswick                   5                  16                      0   \n",
       "Newfoundland and Labrador       0                   0                      0   \n",
       "Nova Scotia                    12                 439                      0   \n",
       "Ontario                      5623                 493                      0   \n",
       "Saskatchewan                 3677                 344                      0   \n",
       "\n",
       "                           coverage_breadth  coverage_depth     Ct  \\\n",
       "province                                                             \n",
       "Alberta                                   0           26123  27740   \n",
       "British Columbia                          0               0  34697   \n",
       "Manitoba                                  0               0    124   \n",
       "New Brunswick                             0               0     90   \n",
       "Newfoundland and Labrador                 0               0      0   \n",
       "Nova Scotia                               0               0    106   \n",
       "Ontario                                 111             672  12022   \n",
       "Saskatchewan                              0              80   2831   \n",
       "\n",
       "                           gisaid_accession  \n",
       "province                                     \n",
       "Alberta                                1704  \n",
       "British Columbia                          0  \n",
       "Manitoba                                  0  \n",
       "New Brunswick                             0  \n",
       "Newfoundland and Labrador                 0  \n",
       "Nova Scotia                               0  \n",
       "Ontario                                 111  \n",
       "Saskatchewan                              0  "
      ]
     },
     "execution_count": 120,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seqs.columns = ['collection_date', 'province', 'sampling_purpose', 'age', 'age_bin', 'gender', 'sequencing_purpose', 'sequencing_instrument',\n",
    "'coverage_breadth', 'coverage_depth', 'Ct', 'gisaid_accession']\n",
    "seqs.groupby('province').count().rsub(seqs.groupby('province').size(), axis = 0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### How many samples are there from each province?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "province\n",
       "Alberta                      27740\n",
       "British Columbia             34697\n",
       "Manitoba                      1021\n",
       "New Brunswick                  507\n",
       "Newfoundland and Labrador      296\n",
       "Nova Scotia                   1202\n",
       "Ontario                      29685\n",
       "Saskatchewan                 10209\n",
       "dtype: int64"
      ]
     },
     "execution_count": 121,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seqs.groupby('province').size()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### How many complete values are there from each province in each field?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>collection_date</th>\n",
       "      <th>sampling_purpose</th>\n",
       "      <th>age</th>\n",
       "      <th>age_bin</th>\n",
       "      <th>gender</th>\n",
       "      <th>sequencing_purpose</th>\n",
       "      <th>sequencing_instrument</th>\n",
       "      <th>coverage_breadth</th>\n",
       "      <th>coverage_depth</th>\n",
       "      <th>Ct</th>\n",
       "      <th>gisaid_accession</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>province</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Alberta</th>\n",
       "      <td>27740</td>\n",
       "      <td>27740</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>27740</td>\n",
       "      <td>27740</td>\n",
       "      <td>27740</td>\n",
       "      <td>1617</td>\n",
       "      <td>0</td>\n",
       "      <td>26036</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>British Columbia</th>\n",
       "      <td>34697</td>\n",
       "      <td>12121</td>\n",
       "      <td>0</td>\n",
       "      <td>12121</td>\n",
       "      <td>12121</td>\n",
       "      <td>12121</td>\n",
       "      <td>34697</td>\n",
       "      <td>34697</td>\n",
       "      <td>34697</td>\n",
       "      <td>0</td>\n",
       "      <td>34697</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Manitoba</th>\n",
       "      <td>1021</td>\n",
       "      <td>1021</td>\n",
       "      <td>999</td>\n",
       "      <td>1021</td>\n",
       "      <td>1021</td>\n",
       "      <td>996</td>\n",
       "      <td>1021</td>\n",
       "      <td>1021</td>\n",
       "      <td>1021</td>\n",
       "      <td>897</td>\n",
       "      <td>1021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>New Brunswick</th>\n",
       "      <td>507</td>\n",
       "      <td>507</td>\n",
       "      <td>0</td>\n",
       "      <td>491</td>\n",
       "      <td>502</td>\n",
       "      <td>491</td>\n",
       "      <td>507</td>\n",
       "      <td>507</td>\n",
       "      <td>507</td>\n",
       "      <td>417</td>\n",
       "      <td>507</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Newfoundland and Labrador</th>\n",
       "      <td>296</td>\n",
       "      <td>296</td>\n",
       "      <td>292</td>\n",
       "      <td>296</td>\n",
       "      <td>296</td>\n",
       "      <td>296</td>\n",
       "      <td>296</td>\n",
       "      <td>296</td>\n",
       "      <td>296</td>\n",
       "      <td>296</td>\n",
       "      <td>296</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Nova Scotia</th>\n",
       "      <td>1202</td>\n",
       "      <td>1202</td>\n",
       "      <td>0</td>\n",
       "      <td>1147</td>\n",
       "      <td>1190</td>\n",
       "      <td>763</td>\n",
       "      <td>1202</td>\n",
       "      <td>1202</td>\n",
       "      <td>1202</td>\n",
       "      <td>1096</td>\n",
       "      <td>1202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ontario</th>\n",
       "      <td>29685</td>\n",
       "      <td>29574</td>\n",
       "      <td>0</td>\n",
       "      <td>26456</td>\n",
       "      <td>24062</td>\n",
       "      <td>29192</td>\n",
       "      <td>29685</td>\n",
       "      <td>29574</td>\n",
       "      <td>29013</td>\n",
       "      <td>17663</td>\n",
       "      <td>29574</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Saskatchewan</th>\n",
       "      <td>10209</td>\n",
       "      <td>10209</td>\n",
       "      <td>6302</td>\n",
       "      <td>6362</td>\n",
       "      <td>6532</td>\n",
       "      <td>9865</td>\n",
       "      <td>10209</td>\n",
       "      <td>10209</td>\n",
       "      <td>10129</td>\n",
       "      <td>7378</td>\n",
       "      <td>10209</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           collection_date  sampling_purpose   age  age_bin  \\\n",
       "province                                                                      \n",
       "Alberta                              27740             27740     0        0   \n",
       "British Columbia                     34697             12121     0    12121   \n",
       "Manitoba                              1021              1021   999     1021   \n",
       "New Brunswick                          507               507     0      491   \n",
       "Newfoundland and Labrador              296               296   292      296   \n",
       "Nova Scotia                           1202              1202     0     1147   \n",
       "Ontario                              29685             29574     0    26456   \n",
       "Saskatchewan                         10209             10209  6302     6362   \n",
       "\n",
       "                           gender  sequencing_purpose  sequencing_instrument  \\\n",
       "province                                                                       \n",
       "Alberta                         0               27740                  27740   \n",
       "British Columbia            12121               12121                  34697   \n",
       "Manitoba                     1021                 996                   1021   \n",
       "New Brunswick                 502                 491                    507   \n",
       "Newfoundland and Labrador     296                 296                    296   \n",
       "Nova Scotia                  1190                 763                   1202   \n",
       "Ontario                     24062               29192                  29685   \n",
       "Saskatchewan                 6532                9865                  10209   \n",
       "\n",
       "                           coverage_breadth  coverage_depth     Ct  \\\n",
       "province                                                             \n",
       "Alberta                               27740            1617      0   \n",
       "British Columbia                      34697           34697      0   \n",
       "Manitoba                               1021            1021    897   \n",
       "New Brunswick                           507             507    417   \n",
       "Newfoundland and Labrador               296             296    296   \n",
       "Nova Scotia                            1202            1202   1096   \n",
       "Ontario                               29574           29013  17663   \n",
       "Saskatchewan                          10209           10129   7378   \n",
       "\n",
       "                           gisaid_accession  \n",
       "province                                     \n",
       "Alberta                               26036  \n",
       "British Columbia                      34697  \n",
       "Manitoba                               1021  \n",
       "New Brunswick                           507  \n",
       "Newfoundland and Labrador               296  \n",
       "Nova Scotia                            1202  \n",
       "Ontario                               29574  \n",
       "Saskatchewan                          10209  "
      ]
     },
     "execution_count": 122,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts = seqs.groupby('province').count()\n",
    "counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [],
   "source": [
    "totals = counts['collection_date']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### What proportion of each field is complete for each province?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>collection_date</th>\n",
       "      <th>sampling_purpose</th>\n",
       "      <th>age</th>\n",
       "      <th>age_bin</th>\n",
       "      <th>gender</th>\n",
       "      <th>sequencing_purpose</th>\n",
       "      <th>sequencing_instrument</th>\n",
       "      <th>coverage_breadth</th>\n",
       "      <th>coverage_depth</th>\n",
       "      <th>Ct</th>\n",
       "      <th>gisaid_accession</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>province</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Alberta</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.058291</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.938572</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>British Columbia</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.349339</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.349339</td>\n",
       "      <td>0.349339</td>\n",
       "      <td>0.349339</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Manitoba</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.978452</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.975514</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.878550</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>New Brunswick</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.968442</td>\n",
       "      <td>0.990138</td>\n",
       "      <td>0.968442</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.822485</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Newfoundland and Labrador</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.986486</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Nova Scotia</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.954243</td>\n",
       "      <td>0.990017</td>\n",
       "      <td>0.634775</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.911814</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ontario</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.996261</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.891225</td>\n",
       "      <td>0.810578</td>\n",
       "      <td>0.983392</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.996261</td>\n",
       "      <td>0.977362</td>\n",
       "      <td>0.595014</td>\n",
       "      <td>0.996261</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Saskatchewan</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.617298</td>\n",
       "      <td>0.623176</td>\n",
       "      <td>0.639828</td>\n",
       "      <td>0.966304</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.992164</td>\n",
       "      <td>0.722696</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           collection_date  sampling_purpose       age  \\\n",
       "province                                                                 \n",
       "Alberta                                1.0          1.000000  0.000000   \n",
       "British Columbia                       1.0          0.349339  0.000000   \n",
       "Manitoba                               1.0          1.000000  0.978452   \n",
       "New Brunswick                          1.0          1.000000  0.000000   \n",
       "Newfoundland and Labrador              1.0          1.000000  0.986486   \n",
       "Nova Scotia                            1.0          1.000000  0.000000   \n",
       "Ontario                                1.0          0.996261  0.000000   \n",
       "Saskatchewan                           1.0          1.000000  0.617298   \n",
       "\n",
       "                            age_bin    gender  sequencing_purpose  \\\n",
       "province                                                            \n",
       "Alberta                    0.000000  0.000000            1.000000   \n",
       "British Columbia           0.349339  0.349339            0.349339   \n",
       "Manitoba                   1.000000  1.000000            0.975514   \n",
       "New Brunswick              0.968442  0.990138            0.968442   \n",
       "Newfoundland and Labrador  1.000000  1.000000            1.000000   \n",
       "Nova Scotia                0.954243  0.990017            0.634775   \n",
       "Ontario                    0.891225  0.810578            0.983392   \n",
       "Saskatchewan               0.623176  0.639828            0.966304   \n",
       "\n",
       "                           sequencing_instrument  coverage_breadth  \\\n",
       "province                                                             \n",
       "Alberta                                      1.0          1.000000   \n",
       "British Columbia                             1.0          1.000000   \n",
       "Manitoba                                     1.0          1.000000   \n",
       "New Brunswick                                1.0          1.000000   \n",
       "Newfoundland and Labrador                    1.0          1.000000   \n",
       "Nova Scotia                                  1.0          1.000000   \n",
       "Ontario                                      1.0          0.996261   \n",
       "Saskatchewan                                 1.0          1.000000   \n",
       "\n",
       "                           coverage_depth        Ct  gisaid_accession  \n",
       "province                                                               \n",
       "Alberta                          0.058291  0.000000          0.938572  \n",
       "British Columbia                 1.000000  0.000000          1.000000  \n",
       "Manitoba                         1.000000  0.878550          1.000000  \n",
       "New Brunswick                    1.000000  0.822485          1.000000  \n",
       "Newfoundland and Labrador        1.000000  1.000000          1.000000  \n",
       "Nova Scotia                      1.000000  0.911814          1.000000  \n",
       "Ontario                          0.977362  0.595014          0.996261  \n",
       "Saskatchewan                     0.992164  0.722696          1.000000  "
      ]
     },
     "execution_count": 124,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seqs_comp = deepcopy(pd.DataFrame(seqs))\n",
    "seqs_comp = seqs_comp.groupby('province').count()\n",
    "\n",
    "seqs_comp.divide(totals, axis = 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>collection_date</th>\n",
       "      <th>province</th>\n",
       "      <th>sampling_purpose</th>\n",
       "      <th>age</th>\n",
       "      <th>age_bin</th>\n",
       "      <th>gender</th>\n",
       "      <th>sequencing_purpose</th>\n",
       "      <th>sequencing_instrument</th>\n",
       "      <th>coverage_breadth</th>\n",
       "      <th>coverage_depth</th>\n",
       "      <th>Ct</th>\n",
       "      <th>gisaid_accession</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2020-12-15</td>\n",
       "      <td>Ontario</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Oxford Nanopore</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5384.4X</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2020-10-12</td>\n",
       "      <td>Ontario</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Oxford Nanopore</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2912.4X</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2020-10-02</td>\n",
       "      <td>Ontario</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Oxford Nanopore</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2661.6X</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2020-04-12</td>\n",
       "      <td>Ontario</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Illumina</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1671.9X</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2020-04-22</td>\n",
       "      <td>Ontario</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Illumina</td>\n",
       "      <td>NaN</td>\n",
       "      <td>959.3X</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  collection_date province sampling_purpose  age age_bin gender  \\\n",
       "0      2020-12-15  Ontario              NaN  NaN     NaN    NaN   \n",
       "1      2020-10-12  Ontario              NaN  NaN     NaN    NaN   \n",
       "2      2020-10-02  Ontario              NaN  NaN     NaN    NaN   \n",
       "3      2020-04-12  Ontario              NaN  NaN     NaN    NaN   \n",
       "4      2020-04-22  Ontario              NaN  NaN     NaN    NaN   \n",
       "\n",
       "  sequencing_purpose sequencing_instrument coverage_breadth coverage_depth  \\\n",
       "0                NaN       Oxford Nanopore              NaN        5384.4X   \n",
       "1                NaN       Oxford Nanopore              NaN        2912.4X   \n",
       "2                NaN       Oxford Nanopore              NaN        2661.6X   \n",
       "3                NaN              Illumina              NaN        1671.9X   \n",
       "4                NaN              Illumina              NaN         959.3X   \n",
       "\n",
       "   Ct gisaid_accession  \n",
       "0 NaN              NaN  \n",
       "1 NaN              NaN  \n",
       "2 NaN              NaN  \n",
       "3 NaN              NaN  \n",
       "4 NaN              NaN  "
      ]
     },
     "execution_count": 125,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seqs.head()"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "40d3a090f54c6569ab1632332b64b2c03c39dcf918b08424e98f38b5ae0af88f"
  },
  "kernelspec": {
   "display_name": "Python 3.8.11 64-bit ('base': conda)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.11"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
