{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Import libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Read the sample data and see what it looks like"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>study_id</th>\n",
       "      <th>specimen collector sample ID</th>\n",
       "      <th>sample collected by</th>\n",
       "      <th>sequence submitted by</th>\n",
       "      <th>sample collection date</th>\n",
       "      <th>sample collection date null reason</th>\n",
       "      <th>geo_loc_name (country)</th>\n",
       "      <th>geo_loc_name (state/province/territory)</th>\n",
       "      <th>organism</th>\n",
       "      <th>isolate</th>\n",
       "      <th>...</th>\n",
       "      <th>consensus sequence software name</th>\n",
       "      <th>consensus sequence software version</th>\n",
       "      <th>breadth of coverage value</th>\n",
       "      <th>depth of coverage value</th>\n",
       "      <th>reference genome accession</th>\n",
       "      <th>bioinformatics protocol</th>\n",
       "      <th>gene name</th>\n",
       "      <th>diagnostic pcr Ct value</th>\n",
       "      <th>diagnostic pcr Ct value null reason</th>\n",
       "      <th>GISAID accession</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>UHTC-ON</td>\n",
       "      <td>UHTC_0471</td>\n",
       "      <td>Unity Health Toronto</td>\n",
       "      <td>Ontario Institute for Cancer Research (OICR)</td>\n",
       "      <td>2020-12-15</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Canada</td>\n",
       "      <td>Ontario</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>hCoV-19/Canada/ON-UHTC_0471/2021</td>\n",
       "      <td>...</td>\n",
       "      <td>ARTIC-nanopolish</td>\n",
       "      <td>1.1.3</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>5384.4X</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>UHTC-ON</td>\n",
       "      <td>UHTC_0464</td>\n",
       "      <td>Unity Health Toronto</td>\n",
       "      <td>Ontario Institute for Cancer Research (OICR)</td>\n",
       "      <td>2020-10-12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Canada</td>\n",
       "      <td>Ontario</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>hCoV-19/Canada/ON-UHTC_0464/2021</td>\n",
       "      <td>...</td>\n",
       "      <td>ARTIC-nanopolish</td>\n",
       "      <td>1.1.3</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>2912.4X</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>UHTC-ON</td>\n",
       "      <td>UHTC_0463</td>\n",
       "      <td>Unity Health Toronto</td>\n",
       "      <td>Ontario Institute for Cancer Research (OICR)</td>\n",
       "      <td>2020-10-02</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Canada</td>\n",
       "      <td>Ontario</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>hCoV-19/Canada/ON-UHTC_0463/2021</td>\n",
       "      <td>...</td>\n",
       "      <td>ARTIC-nanopolish</td>\n",
       "      <td>1.1.3</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>2661.6X</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>KHSC-ON</td>\n",
       "      <td>SLB3076</td>\n",
       "      <td>Queen's University / Kingston Health Sciences ...</td>\n",
       "      <td>Ontario Institute for Cancer Research (OICR)</td>\n",
       "      <td>2020-04-12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Canada</td>\n",
       "      <td>Ontario</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>hCoV-19/Canada/ON-SLB3076/2021</td>\n",
       "      <td>...</td>\n",
       "      <td>ncov2019-artic-nf</td>\n",
       "      <td>OICR v1.6</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>1671.9X</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>KHSC-ON</td>\n",
       "      <td>SLB3121</td>\n",
       "      <td>Queen's University / Kingston Health Sciences ...</td>\n",
       "      <td>Ontario Institute for Cancer Research (OICR)</td>\n",
       "      <td>2020-04-22</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Canada</td>\n",
       "      <td>Ontario</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>hCoV-19/Canada/ON-SLB3121/2021</td>\n",
       "      <td>...</td>\n",
       "      <td>ncov2019-artic-nf</td>\n",
       "      <td>OICR v1.6</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>959.3X</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 43 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  study_id specimen collector sample ID  \\\n",
       "0  UHTC-ON                    UHTC_0471   \n",
       "1  UHTC-ON                    UHTC_0464   \n",
       "2  UHTC-ON                    UHTC_0463   \n",
       "3  KHSC-ON                      SLB3076   \n",
       "4  KHSC-ON                      SLB3121   \n",
       "\n",
       "                                 sample collected by  \\\n",
       "0                               Unity Health Toronto   \n",
       "1                               Unity Health Toronto   \n",
       "2                               Unity Health Toronto   \n",
       "3  Queen's University / Kingston Health Sciences ...   \n",
       "4  Queen's University / Kingston Health Sciences ...   \n",
       "\n",
       "                          sequence submitted by sample collection date  \\\n",
       "0  Ontario Institute for Cancer Research (OICR)             2020-12-15   \n",
       "1  Ontario Institute for Cancer Research (OICR)             2020-10-12   \n",
       "2  Ontario Institute for Cancer Research (OICR)             2020-10-02   \n",
       "3  Ontario Institute for Cancer Research (OICR)             2020-04-12   \n",
       "4  Ontario Institute for Cancer Research (OICR)             2020-04-22   \n",
       "\n",
       "   sample collection date null reason geo_loc_name (country)  \\\n",
       "0                                 NaN                 Canada   \n",
       "1                                 NaN                 Canada   \n",
       "2                                 NaN                 Canada   \n",
       "3                                 NaN                 Canada   \n",
       "4                                 NaN                 Canada   \n",
       "\n",
       "  geo_loc_name (state/province/territory)      organism  \\\n",
       "0                                 Ontario  Not Provided   \n",
       "1                                 Ontario  Not Provided   \n",
       "2                                 Ontario  Not Provided   \n",
       "3                                 Ontario  Not Provided   \n",
       "4                                 Ontario  Not Provided   \n",
       "\n",
       "                            isolate  ... consensus sequence software name  \\\n",
       "0  hCoV-19/Canada/ON-UHTC_0471/2021  ...                 ARTIC-nanopolish   \n",
       "1  hCoV-19/Canada/ON-UHTC_0464/2021  ...                 ARTIC-nanopolish   \n",
       "2  hCoV-19/Canada/ON-UHTC_0463/2021  ...                 ARTIC-nanopolish   \n",
       "3    hCoV-19/Canada/ON-SLB3076/2021  ...                ncov2019-artic-nf   \n",
       "4    hCoV-19/Canada/ON-SLB3121/2021  ...                ncov2019-artic-nf   \n",
       "\n",
       "  consensus sequence software version breadth of coverage value  \\\n",
       "0                               1.1.3              Not Provided   \n",
       "1                               1.1.3              Not Provided   \n",
       "2                               1.1.3              Not Provided   \n",
       "3                           OICR v1.6              Not Provided   \n",
       "4                           OICR v1.6              Not Provided   \n",
       "\n",
       "  depth of coverage value reference genome accession bioinformatics protocol  \\\n",
       "0                 5384.4X                        NaN                     NaN   \n",
       "1                 2912.4X                        NaN                     NaN   \n",
       "2                 2661.6X                        NaN                     NaN   \n",
       "3                 1671.9X                        NaN                     NaN   \n",
       "4                  959.3X                        NaN                     NaN   \n",
       "\n",
       "      gene name diagnostic pcr Ct value diagnostic pcr Ct value null reason  \\\n",
       "0  Not Provided                     NaN                        Not Provided   \n",
       "1  Not Provided                     NaN                        Not Provided   \n",
       "2  Not Provided                     NaN                        Not Provided   \n",
       "3  Not Provided                     NaN                        Not Provided   \n",
       "4  Not Provided                     NaN                        Not Provided   \n",
       "\n",
       "  GISAID accession  \n",
       "0              NaN  \n",
       "1              NaN  \n",
       "2              NaN  \n",
       "3              NaN  \n",
       "4              NaN  \n",
       "\n",
       "[5 rows x 43 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seqs = pd.read_csv(\"data/files-archive.tsv\", sep = \"\\t\")\n",
    "seqs.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Print out all column names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['study_id', 'specimen collector sample ID', 'sample collected by',\n",
      "       'sequence submitted by', 'sample collection date',\n",
      "       'sample collection date null reason', 'geo_loc_name (country)',\n",
      "       'geo_loc_name (state/province/territory)', 'organism', 'isolate',\n",
      "       'fasta header name', 'purpose of sampling',\n",
      "       'purpose of sampling details', 'anatomical material', 'anatomical part',\n",
      "       'body product', 'environmental material', 'environmental site',\n",
      "       'collection device', 'collection method', 'host (scientific name)',\n",
      "       'host disease', 'host age', 'host age null reason', 'host age unit',\n",
      "       'host age bin', 'host gender', 'purpose of sequencing',\n",
      "       'purpose of sequencing details', 'sequencing instrument',\n",
      "       'sequencing protocol', 'raw sequence data processing method',\n",
      "       'dehosting method', 'consensus sequence software name',\n",
      "       'consensus sequence software version', 'breadth of coverage value',\n",
      "       'depth of coverage value', 'reference genome accession',\n",
      "       'bioinformatics protocol', 'gene name', 'diagnostic pcr Ct value',\n",
      "       'diagnostic pcr Ct value null reason', 'GISAID accession'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "print(seqs.columns)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Remove the columns we don't need"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "seqs.drop(columns=['study_id', 'specimen collector sample ID', 'sample collected by',\n",
    "       'sequence submitted by',\n",
    "       'sample collection date null reason', 'geo_loc_name (country)',\n",
    "       'organism', 'isolate',\n",
    "       'fasta header name',\n",
    "       'purpose of sampling details', 'anatomical material', 'anatomical part',\n",
    "       'body product', 'environmental material', 'environmental site',\n",
    "       'collection device', 'collection method', 'host (scientific name)',\n",
    "       'host disease','host age null reason', 'host age unit',\n",
    "       'purpose of sequencing details',\n",
    "       'sequencing protocol', 'raw sequence data processing method',\n",
    "       'dehosting method', 'consensus sequence software name',\n",
    "       'consensus sequence software version', 'reference genome accession',\n",
    "       'bioinformatics protocol', 'gene name', \n",
    "       'diagnostic pcr Ct value null reason'], inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample collection date</th>\n",
       "      <th>geo_loc_name (state/province/territory)</th>\n",
       "      <th>purpose of sampling</th>\n",
       "      <th>host age</th>\n",
       "      <th>host age bin</th>\n",
       "      <th>host gender</th>\n",
       "      <th>purpose of sequencing</th>\n",
       "      <th>sequencing instrument</th>\n",
       "      <th>breadth of coverage value</th>\n",
       "      <th>depth of coverage value</th>\n",
       "      <th>diagnostic pcr Ct value</th>\n",
       "      <th>GISAID accession</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2020-12-15</td>\n",
       "      <td>Ontario</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>Oxford Nanopore</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>5384.4X</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2020-10-12</td>\n",
       "      <td>Ontario</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>Oxford Nanopore</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>2912.4X</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2020-10-02</td>\n",
       "      <td>Ontario</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>Oxford Nanopore</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>2661.6X</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2020-04-12</td>\n",
       "      <td>Ontario</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>Illumina</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>1671.9X</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2020-04-22</td>\n",
       "      <td>Ontario</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>Illumina</td>\n",
       "      <td>Not Provided</td>\n",
       "      <td>959.3X</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample collection date geo_loc_name (state/province/territory)  \\\n",
       "0             2020-12-15                                 Ontario   \n",
       "1             2020-10-12                                 Ontario   \n",
       "2             2020-10-02                                 Ontario   \n",
       "3             2020-04-12                                 Ontario   \n",
       "4             2020-04-22                                 Ontario   \n",
       "\n",
       "  purpose of sampling  host age    host age bin   host gender  \\\n",
       "0        Not Provided       NaN  Not Applicable  Not Provided   \n",
       "1        Not Provided       NaN  Not Applicable  Not Provided   \n",
       "2        Not Provided       NaN  Not Applicable  Not Provided   \n",
       "3        Not Provided       NaN  Not Applicable  Not Provided   \n",
       "4        Not Provided       NaN  Not Applicable  Not Provided   \n",
       "\n",
       "  purpose of sequencing sequencing instrument breadth of coverage value  \\\n",
       "0          Not Provided       Oxford Nanopore              Not Provided   \n",
       "1          Not Provided       Oxford Nanopore              Not Provided   \n",
       "2          Not Provided       Oxford Nanopore              Not Provided   \n",
       "3          Not Provided              Illumina              Not Provided   \n",
       "4          Not Provided              Illumina              Not Provided   \n",
       "\n",
       "  depth of coverage value  diagnostic pcr Ct value GISAID accession  \n",
       "0                 5384.4X                      NaN              NaN  \n",
       "1                 2912.4X                      NaN              NaN  \n",
       "2                 2661.6X                      NaN              NaN  \n",
       "3                 1671.9X                      NaN              NaN  \n",
       "4                  959.3X                      NaN              NaN  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seqs.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Let's replace the many forms of 'Not Provided' or NaN with np.nan and convert collection date to a date time object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample collection date</th>\n",
       "      <th>geo_loc_name (state/province/territory)</th>\n",
       "      <th>purpose of sampling</th>\n",
       "      <th>host age</th>\n",
       "      <th>host age bin</th>\n",
       "      <th>host gender</th>\n",
       "      <th>purpose of sequencing</th>\n",
       "      <th>sequencing instrument</th>\n",
       "      <th>breadth of coverage value</th>\n",
       "      <th>depth of coverage value</th>\n",
       "      <th>diagnostic pcr Ct value</th>\n",
       "      <th>GISAID accession</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2020-12-15</td>\n",
       "      <td>Ontario</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Oxford Nanopore</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5384.4X</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2020-10-12</td>\n",
       "      <td>Ontario</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Oxford Nanopore</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2912.4X</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2020-10-02</td>\n",
       "      <td>Ontario</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Oxford Nanopore</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2661.6X</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2020-04-12</td>\n",
       "      <td>Ontario</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Illumina</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1671.9X</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2020-04-22</td>\n",
       "      <td>Ontario</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Illumina</td>\n",
       "      <td>NaN</td>\n",
       "      <td>959.3X</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample collection date geo_loc_name (state/province/territory)  \\\n",
       "0             2020-12-15                                 Ontario   \n",
       "1             2020-10-12                                 Ontario   \n",
       "2             2020-10-02                                 Ontario   \n",
       "3             2020-04-12                                 Ontario   \n",
       "4             2020-04-22                                 Ontario   \n",
       "\n",
       "  purpose of sampling  host age host age bin host gender  \\\n",
       "0                 NaN       NaN          NaN         NaN   \n",
       "1                 NaN       NaN          NaN         NaN   \n",
       "2                 NaN       NaN          NaN         NaN   \n",
       "3                 NaN       NaN          NaN         NaN   \n",
       "4                 NaN       NaN          NaN         NaN   \n",
       "\n",
       "  purpose of sequencing sequencing instrument breadth of coverage value  \\\n",
       "0                   NaN       Oxford Nanopore                       NaN   \n",
       "1                   NaN       Oxford Nanopore                       NaN   \n",
       "2                   NaN       Oxford Nanopore                       NaN   \n",
       "3                   NaN              Illumina                       NaN   \n",
       "4                   NaN              Illumina                       NaN   \n",
       "\n",
       "  depth of coverage value  diagnostic pcr Ct value GISAID accession  \n",
       "0                 5384.4X                      NaN              NaN  \n",
       "1                 2912.4X                      NaN              NaN  \n",
       "2                 2661.6X                      NaN              NaN  \n",
       "3                 1671.9X                      NaN              NaN  \n",
       "4                  959.3X                      NaN              NaN  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seqs.replace({'Not Provided':np.nan, 'NaN':np.nan, 'Not Applicable':np.nan}, inplace = True)\n",
    "seqs['sample collection date'] = pd.to_datetime(seqs['sample collection date'])\n",
    "seqs.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### What percent of the data is missing?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "sample collection date                      0.000000\n",
       "geo_loc_name (state/province/territory)     0.000000\n",
       "purpose of sampling                        21.533453\n",
       "host age                                   92.793075\n",
       "host age bin                               54.541226\n",
       "host gender                                56.600890\n",
       "purpose of sequencing                      22.678132\n",
       "sequencing instrument                       0.000000\n",
       "breadth of coverage value                   0.105356\n",
       "depth of coverage value                    25.508509\n",
       "diagnostic pcr Ct value                    73.663829\n",
       "GISAID accession                            1.722714\n",
       "dtype: float64"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seqs.isna().sum()/len(seqs)*100"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### All samples have associated collection dates!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "40d3a090f54c6569ab1632332b64b2c03c39dcf918b08424e98f38b5ae0af88f"
  },
  "kernelspec": {
   "display_name": "Python 3.8.11 64-bit ('base': conda)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.11"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
